### PR TITLE
[WIP] feat: Rewrite to automation-extra, Support both Playwright and Puppeteer

### PR DESCRIPTION
**Pull request by berstend**
### Update: [Beta versions are available for testing now](https://github.com/berstend/puppeteer-extra/issues/454)

---

### Gist

- massive rewrite of the whole project 😅 
- adds a new `automation-extra` package, which is the underlying shared foundation for both `playwright-extra` and the new `puppeteer-extra` package
- adds a new `automation-extra-plugin` (cleaner API, adds Playwright events, better type safety)
- many more things, e.g. Plugins don't listen to events themselves anymore but get called from the mother ship (we had a set of issues with e.g. too many listeners that should be fixed by that)
- much improved type safety all around

Motivation for adding playwright support: https://github.com/puppeteer/puppeteer/issues/3667

### Status

- `automation-extra` - Finished: We have plugin support for both Playwright & Puppeteer
- `playwright-extra` - Finished: It's just a small entry point to `automation-extra` to make things more clear & type safe
- `automation-extra-plugin` - Finished for now: ~~I want to add shims to make plugins supporting both pptr & playwright less verbose (e.g. pptr has `page.evaluateOnNewDocument`, playwright has `context.addInitScript`)~~

~~I haven't switched `puppeteer-extra` over to use the new `automation-extra` code base yet, but specific tests against `stealth`, `adblocker`, `recaptcha` are all green 🚀~~

### Next steps

I realized pretty late in the process of adding Playwright support that I need to make the code backwards compatible with existing plugins, so new `automation-extra-plugin` based plugins work with both `playwright-extra` and `puppeteer-extra` (the latter still supporting `puppeteer-extra-plugin`'s). 😅

- [x] ~~Replace `puppeteer-extra` with a small package using the new code in `automation-extra`~~
- [x] ~~Make sure all tests in the whole monorepo are green after that~~
- [x] ~~Port the first plugin over to support both Playwright & Puppeteer~~
- [x] ~~Add/update/improve/fix documentation where needed~~
- [ ] Final public beta version testing before release

Once that's done we should be able to push/publish this stuff without breaking anything. Afterwards I'll start porting other existing plugins to support both Playwright and Puppeteer (their APIs differ, so existing puppeteer plugins need to be modified to support both - this PR is about the underlying plugin framework itself).

In the process of porting the plugins I will extend `automation-extra-plugin` wherever it makes sense (add new unified events (e.g. `addScript`) or add some sort of shim).

### Known issues

- [x] ~~I removed that crusty "data from plugins" stuff, so this breaks e.g. the `puppeteer-extra-plugin-user-data-dir` plugin (I'll come up with a cleaner replacement API)~~
- Playwright's `.launchPersistentContext()` is not yet augmented with plugin functionality

### CODE REVIEW

**Comment by XBeg9** [comment]
Wow! That's huge PR @berstend, thanks for playwright support!

**Comment by dilame** [comment]
I'm waiting for this PR as mom waits for her son from the army!!!

**Comment by berstend** [comment]
> I'm waiting for this PR as mom waits for her son from the army!!!

Haha 😄 Unfortunately progress abruptly stopped due to an injury but I'm getting back to it now. :)

**Comment by Nisthar** [comment]
@berstend is it possible to use it now? I am getting "package not found" when trying to install

**Comment by berstend** [comment]
> @berstend is it possible to use it now? I am getting "package not found" when trying to install

Nope, not ready to be used yet. :-)

**Comment by alvis** [comment]
The number of mums waiting for his son back from the army is growing 🤣

**Comment by berstend** [comment]
My apologies, some $work related things got in the way 😅

**Comment by apkoponen** [comment]
I'm also a mother waiting for his son back from the army! 👋  

Is there something we moms here at the home front could do in order to help the war efforts?

**Comment by dilame** [comment]
Join the `Mothers Association` i just created.

**Comment by berstend** [comment]
Haha, this mother meme escalated quickly 😆

Quick update in this regard:

I'm currently working on improved option support in the plugin framework for "umbrella" plugins like `stealth`. Currently if you wish to override e.g. the webgl data you'd need to delete/init/add the `webgl.vendor` evasion plugin from stealth, which is not very ergonomic to put it mildly.

Also options are currently not type safe (aka autocomplete in VSCode), which leaves a lot to be desired.

My goal is something like this (the StealthPlugin itself imports pluginA and pluginB):
```js
const stealth = new StealthPlugin({
  platform: 'MacIntel'
  userAgent: "foobar/1.0",
  webgl: {
    renderer: "RTX 2080",
    vendor: "Intel Inc."
  }
})
```

Where pluginA exposes `platform` & `userAgent` and pluginB exposes `webgl` as options.
I want this dynamic options object to be type safe as well, which is a bit tricky but possible.

As this PR will trigger a major version bump the timing seems right to overhaul the way plugin options are provided as well.

**Comment by berstend** [comment]
Urgh, someone just pushed an unrelated npm package as `playwright-extra`: 
https://github.com/shirshak55/Bee-Connect/issues/39

**Comment by berstend** [comment]
I switched over `puppeteer-extra` to use the new code (in my local branch) and the tests are looking good:

image

I'm gonna port the recaptcha plugin first, to iron out multi-browser tests and to get a feel for which puppeteer/playwright related shims make sense.

**Comment by berstend** [comment]
image

**Comment by berstend** [comment]
First reCAPTCHA solved in Playwright + Firefox + new recaptcha plugin 🎉 

image

Browser support is looking good as well:

image

Only puppeteer + firefox will be an issue, their implementation is super buggy and far from complete (even `evaluateOnNewDocument ` [is missing](https://github.com/puppeteer/puppeteer/issues/6163)). 😅

**Comment by dilame** [comment]
I already set this PR page as my homepage instead of useless `google.com`

**Comment by C2BB** [comment]
Anxiously awaiting! :-) - The best day this time of year: Release Day!!! :-D

**Comment by C2BB** [comment]
Is the automation-extra branch in a safe state to begin testing/use?

**Comment by berstend** [comment]
> Is the automation-extra branch in a safe state to begin testing/use?

Nope :-) Just wait a few days more

**Comment by C2BB** [comment]
> > Is the automation-extra branch in a safe state to begin testing/use?
> 
> Nope :-) Just wait a few days more (finishing up things right now)

Awesome! Thank you for the update.

**Comment by C2BB** [comment]
Curious, is this major update still underway or has the plan changed? Can the community help in any way? Thanks!

**Comment by SiroDiaz** [comment]
Hoping to see this new big update! @berstend, when will you release this update?

**Comment by berstend** [comment]
> when will you release this update?

This rewrite incl. Playwright support is basically finished. :-)

I want to do some final in-depth manual testing with more exotic plugins (`user-data-dir` and the like), but so far everything works. This rewrite is a massive change though, while trying to stay compatible with all existing plugins, so it's better to err on the side of caution. :-)

**The main reason this isn't released yet:** I wasn't sure if it makes sense to release Playwright support before a playwright-supporting stealth plugin is available to accompany it, as it's the most popular plugin. As mentioned in the PR description, even if the plugin framework supports multiple drivers there are still differences which require new/updated plugins.

My intention back in December was to implement a quick chromium-only playwright port of the stealth plugin before release, but then my new year vacation got in the way. :-) 

A new recaptcha plugin with playwright support (all browsers) is already part of this PR and I'm now leaning towards releasing the code before a stealth plugin is available (to finally make the switch 😅).

File: packages/automation-extra/src/base.ts [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+    })(browser.newContext, browser)
+
+    browser.newPage = ((originalMethod, ctx) => {
+      return async (options: pw.BrowserContextOptions = {}) => {
+        const contextOptions: pw.BrowserContextOptions =
+          (await this.plugins.dispatchBlocking(
+            'beforeContext',
+            options,
+            browser
+          )) || options
```
Comment (alec-mccormick) [inline packages/automation-extra/src/base.ts:230]: I believe this dispatch call is a bug. Playwright's `browser.newPage()` internally calls `browser.newContext()` then `context.newPage()` so this is firing the `beforeContext` event twice for every `browser.newPage()` call. [Here's a link to the relevant file & line number in Playwright.](https://github.com/microsoft/playwright/blob/master/src/client/browser.ts#L67)

Comment (alec-mccormick) [reply packages/automation-extra/src/base.ts:230]: Overall, this is an excellent refactor! I have been using this branch quite a lot while working with Playwright and just started noticing a few bugs here and there.

Comment (berstend) [reply packages/automation-extra/src/base.ts:230]: Very interesting, thanks! Will do some local testing around the `beforeContext` dispatch, it's possible for the second call to be redundant.

> Overall, this is an excellent refactor! I have been using this branch quite a lot while working with Playwright and just started noticing a few bugs here and there.

As you're the first brave alpha user of `automation-extra` (😅) kindly mention any bug you might encounter here or on discord, thanks :-)

**Comment by Embry-Bit** [comment]
Excited to see this come out, thank you for your hard work.

**Comment by berstend** [comment]
Unfortunately puppeteer broke TypeScript support in their new v7 release. :-/

image

They decided to ship types with their main package (which is good, playwright does that as well) but unfortunately did so in a broken way (and `@types/puppeteer` is now outdated). Feels a bit weird that such a popular library was released in a broken state like this.

Please upvote this issue so they resolve it quickly:
https://github.com/puppeteer/puppeteer/issues/6816

File: packages/automation-extra/src/base.ts (hunk 1 above)
Comment (berstend) [reply packages/automation-extra/src/base.ts]: @alec-mccormick You were correct, the event was fired twice. :-) The tests didn't catch that as they only checked if the event was fired, not how often. Fix & updated tests: [`4bad9e6` (#303)](https://github.com/berstend/puppeteer-extra/pull/303/commits/4bad9e6daca6cfdede5de230e37c84b747512f33)

Thanks!

**Comment by berstend** [comment]
## Beta versions available :tada:

I'm publishing beta versions with a `next` tag so everyone can help me test for bugs before the public release. 😄 

### `playwright-extra`

```
yarn add playwright playwright-extra@next
```

Currently only the new `@extra/recaptcha` plugin is available (works with Chromium, Firefox, Safari):
```
yarn add @extra/recaptcha@next
```

Documentation & demo code (note, this points to the `automation-extra` branch):
- [playwright-extra](https://github.com/berstend/puppeteer-extra/tree/automation-extra/packages/playwright-extra#readme)
- [@extra/recaptcha](https://github.com/berstend/puppeteer-extra/tree/automation-extra/packages/plugin-recaptcha#readme)

### `puppeteer-extra`

A complete rewrite of the previous package, using the same shared foundation as `playwright-extra`. The goal is to stay 100% compatible with existing code and all `puppeteer-extra-plugin-*` plugins, as well as supporting the new `@extra/*` plugins.

```
yarn add puppeteer puppeteer-extra@next
```

You can (and should be able to) use the new `@extra/recaptcha` with it (`yarn add @extra/recaptcha@next`), as well as all existing `puppeteer-extra` plugins.

## Things to test

- Does `playwright-extra` work well, does it work with all browsers
- Does `@extra/recaptcha` (the first plugin to support Playwright, Puppeteer + all browsers) work well in both `playwright-extra` as well as the new `puppeteer-extra`
- Does the new `puppeteer-extra` work well with existing code bases
- Does the new `puppeteer-extra` work well with existing plugins (especially the less popular ones)
- **TypeScript:** Are there any issues with typings in any of the new packages
- Are there errors in the new end-user documentation or things that should be updated/improved
- In case you're developing plugins: Is the new [`automation-extra-plugin` documentation](https://github.com/berstend/puppeteer-extra/tree/automation-extra/packages/automation-extra-plugin) sufficient
- Any random quirks or odd behavior you might notice

### Notes
- Puppeteer broke typedef support in their latest releases, use `puppeteer@5` when using TypeScript
- Playwright's `.launchPersistentContext()` is not yet augmented with plugin functionality
- `@extra/stealth` with playwright support is not available yet

Let's make mom not wait [any longer](https://github.com/berstend/puppeteer-extra/pull/303#issuecomment-700664383)! 😄

**Comment by berstend** [comment]
**New plugin available to test: `@extra/humanize`**

> A plugin for playwright-extra & puppeteer-extra to humanize input (mouse movements, etc)

Readme:
https://github.com/berstend/puppeteer-extra/tree/automation-extra/packages/plugin-humanize

Install (beta version):
```
yarn add @extra/humanize@next
```

**Comment by opahopa** [comment]
Hi!

i can see that the docs were updated to include `playwright-extra` usage, however the package from npm contains `throw new Error("Coming soon: https://github.com/berstend/puppeteer-extra/pull/303")` is it ready to be used for test purposes or that was just a docs update?

**Comment by berstend** [comment]
@opahopa you need to install `@next` versions currently as mentioned here:
https://github.com/berstend/puppeteer-extra/pull/303#issuecomment-775277480

**Comment by josephlevenson** [comment]
Thank you so much, this is amazing work! Are there any plans for porting stealth over to Playwright in the near future?

**Comment by wesley-campbell-q2** [comment]
@berstend Excellent job on porting this to Playwright. I'm running into an issue when using the recaptcha plugin if you use all the browsers { chromium, firefox, webkit } = require('playwright-extra') it fails to find the executable for any of the browsers.

**Comment by j3lev** [comment]
@wesley-campbell-q2 did you install `playwright` alongside `playwright-extra`?

**Comment by wesley-campbell-q2** [comment]
@j3lev I ran npm install playwright playwright-extra @extra/recaptcha.

**Comment by j3lev** [comment]
@wesley-campbell-q2 Playwright just released 1.9 today and might not work properly with this port yet, try `npm install playwright@1.8`

**Comment by wesley-campbell-q2** [comment]
@j3lev Based off what I'm seeing you need to install as npm i playwright playwright-extra@next but that still fails me

**Comment by lg** [comment]
same issue -- it looks like its downloading some versions of the browsers, axing them, and then downloading new ones.

**Comment by berstend** [comment]
Everyone willing to test the new versions:
- Please follow the [instructions mentioned here](https://github.com/berstend/puppeteer-extra/pull/303#issuecomment-775277480)

That means:
- Installing versions with a `@next` tag (the package READMEs are written for release, when that won't be necessary anymore)
- Please use `yarn`, the canary release flow is a bit wonky and there might be issues with npm

**Comment by berstend** [comment]
> Thank you so much, this is amazing work! Are there any plans for porting stealth over to Playwright in the near future?

I started work on `@extra/stealth` but it's quite a big undertaking with a fair bit of busy work (just look at e.g. all the evasion tests that need to be modified to test both pptr & playwright, etc) :-)

**Comment by j3lev** [comment]
How will playwright updates work? I'm assuming every release will be tied to a specific playwright version?

**Comment by berstend** [comment]
> How will playwright updates work? I'm assuming every release will be tied to a specific playwright version?

The same as with the existing `puppeteer-extra`: You install your own favorite puppeteer/playwright package version and the `-extra` package will use that (it's just a wrapper around those packages to add plugin lifecycle events). :-)

Regarding `@extra/stealth` (and speaking as a plugin framework developer):
puppeteer has shown to be a lot easier to hack on and work with than playwright - the latter switched to use their own intermediate wire protocol instead of using/exposing CDP (chrome devtools protocol) directly, which makes it harder to find solutions for the more advanced stuff that we fixed through raw CDP so far.

**Comment by DrewDaPilot** [comment]
Has the API changed for the humanize plugin? It does not appear to work and the syntax "firefox.use" throws a type error when using playwright. Maybe I am just oblivious but  any pointers would be much appreciated.

**Comment by berstend** [comment]
> Has the API changed for the humanize plugin? It does not appear to work and the syntax "firefox.use" throws a type error when using playwright. Maybe I am just oblivious but any pointers would be much appreciated.

Are you following the instructions [mentioned here](https://github.com/berstend/puppeteer-extra/pull/303#issuecomment-786869671)?

The readme's in the `automation-extra` branch are written from the perspective of being released, so they don't mention the current need to use a `@next` tag when installing the packages (for humanize: `yarn add @extra/humanize@next`).

If you or someone else have issues: Please mention how you installed packages (the exact command line)

**Comment by berstend** [comment]
I created a canonical issue with condensed info on how to install the new `playwright-extra` & `puppeteer-extra` beta versions and new plugins: https://github.com/berstend/puppeteer-extra/issues/454

**Comment by j3lev** [comment]
Hey @berstend! There is an issue when using versions of Playwright greater than 1.8.0. I ran into this when attempting to using Playwright 1.10.0 with playwright-extra inside a docker container. The browser launch fails because the library tries to use the 1.8 browser binaries. I suspect this might have something to do with the version being locked here https://github.com/berstend/puppeteer-extra/blob/cb586077a848241f119b3a9c051e93babc2ce7a8/packages/playwright-extra/package.json#L64, are there any plans to support later versions of Playwright?

**Comment by berstend** [comment]
I locked this PR so subscribers only get pinged when there's official "updates" of sorts. :-)

For feedback and bug reports on the new automation-extra branch please use this ticket: https://github.com/berstend/puppeteer-extra/issues/454

We're really close to finalizing the new code and the release should happen soon.

**Comment by berstend** [comment]
`playwright-extra` has landed in #664 😄 
readme: https://github.com/berstend/puppeteer-extra/tree/master/packages/playwright-extra

I'm closing this PR/rewrite for now, it's a good reference for future additions though.

**Pull request closed without merging**